posts/2020-03-10-calculating-pi-concurrently.adoc

```diff
@@ -0,0 +1,917 @@
+Started @ offset: 2491
+Started @ offset: 2538
+Started @ offset: 2585
+Started @ offset: 2632
+Started @ offset: 2679
+Started @ offset: 2726
+Started @ offset: 2773
+Started @ offset: 2820
+Started @ offset: 2867
+Started @ offset: 2914
+Started @ offset: 2961
+Pi:      3.14159265358979323846264338327950288419716939937510582097494459230781640628...
+|Delta|: 2.8076968E-3630
+Calculation time: 3,560
+----
+
+What we can observe is that, with 3,008 terms, we have correctly calculated more
+than 3,600 digits accurately.
+
+If we set the size of the ForkJoinPool to 1, 2, 4, 8, 16, 32 we get the
+following values for the calculation time (in seconds) respectively: 18.25,
+9.23, 5.91, 4.02, 4.47, 5.86
+
+So, we see a near linear speed-up by going from one thread to two threads.
+A further increase of the thread-count doesn't yield a further linear speed-up:
+this may be caused by different factors, not in the least by the fact that we
+have a single chip processor with a shared on-chip cache.
+Of course, since it's a four core CPU (with hyper-threads that don't yield the
+same performance as the regular CPU threads), we don't get a speed-up beyond 8
+threads in the ForkJoinPool.
+
+== Conclusion on the `Future` approach
+
+Scala's `Future` API presents a very powerful way to perform asynchronous and
+concurrent execution of code. Even though it may take some time to get one's
+head around, once you grasp it, it's pretty cool and very powerful.
+
+Now, as for π, is approach used in this article to calculate π a realistic way
+to calculate this number to say multi-billion digit precision? Not really for
+multiple reasons.
+
+First of all, this algorithm runs in-memory. If we consider that the current
+record for calculating π digits is at 50 trillion digits, there's no computer
+that can hold the required size of numbers in memory.
+
+Secondly, the algorithm is too slow compared to the algorithm that is currently
+used to set the record(s).
+
+Consider that the current record holder is _Timothy Mullican_
+who calculated the 50.000.000.000.000 first digits of π. It took 303 days
+to complete the calculation.
+
+What should be mentioned though is that the code that was used to perform the
+calculation is http://www.numberworld.org/y-cruncher[y-cruncher]. This C++
+program has been written by Alexander Yee. Between 2010 and 2013, he and
+Shigeru Kondo set various records for calculating π. If you're interested,
+have a look at http://www.numberworld.org/misc_runs/pi-12t[this website].
+
+_y-cruncher_ utilises a different formula than the one used in this article,
+namely the https://en.wikipedia.org/wiki/Chudnovsky_algorithm[Chudnovsky formula].
+Interesting to note is that earlier work by the famous, self-taught, and brilliant
+Indian mathematician https://en.wikipedia.org/wiki/Srinivasa_Ramanujan[Srinivasa Ramanujan]
+inspired the https://en.wikipedia.org/wiki/Chudnovsky_brothers[Chudnovsky brothers] to come up with their formula.
+
+
```

Comment:
I'm not sure if extra spaces render extra spaces in the produced html or not, but if so you may want to remove this one.